### PR TITLE
Make kwiver::vital::unknown_metadata_item more coherent

### PR DESCRIPTION
**Pull request by daniel-riehm**
The existing definition of `kwiver::vital:unknown_metadata_item` has a number of problems: it seems to try to override non-virtual member functions, silently returns a possibly valid value (0) for `to_double()` and `to_uint64()` when the purpose of the class is that it inherently has no valid value, and for some reason masks the existing non-virtual `to_uint64()` method with one that returns a `double`. This hopes to fix those problems by deleting code and using a unique dummy `unknown_t` class (not convertible to or useful for anything) as the internal type to provoke runtime errors when a cast is attempted.

### CODE REVIEW

**Comment by daniel-riehm** [comment]
@russelldj This may be relevant to #1348.

File: vital/types/metadata.h
Comment (mwoehlke-kitware) [inline vital/types/metadata.h:214]: It seems this should probably be `private`?

**Comment by daniel-riehm** [comment]
Can this land?

File: vital/types/metadata.h
Comment (russelldj) [inline vital/types/metadata.h:213]: I'm fine with changing the name but curious about your rationale?

**Comment by russelldj** [comment]
These changes look fine to me but I haven't gotten a chance to check for issues elsewhere in the codebase. I'd assume most other than changing the `UNKNOWN` string would be caught at compile time.

**Review by russelldj — APPROVED** [review]

File: python/kwiver/vital/tests/test_metadata.py
Comment (tao558) [inline python/kwiver/vital/tests/test_metadata.py:237]: Does `inst.data()` not cause an error here? Seems like we're returning an instance of `unknown_t`, which is not bound

**Review by tao558 — CHANGES_REQUESTED** [review]

File: python/kwiver/vital/tests/test_metadata.py
Comment (daniel-riehm) [reply python/kwiver/vital/tests/test_metadata.py:237]: Yes, that's the reason for adding the check. 

Would it be better perhaps to somehow bind the C++ `unknown_t` to Python's `None`, if possible? Or keep it this way? Or something else...

Comment (tao558) [reply python/kwiver/vital/tests/test_metadata.py:237]: I think treating `unknown_t` as `None` is reasonable on the Python side. I haven't used this class, and I wrote the bindings for it a while ago, however. @mwoehlke-kitware, does this sound reasonable to you?

Comment (daniel-riehm) [reply python/kwiver/vital/tests/test_metadata.py]: I've changed it to `None` on the python side.

Comment (tao558) [reply python/kwiver/vital/tests/test_metadata.py]: `type()` would also need to change. I'd like to get approval from either @mwoehlke-kitware or @russelldj before making the switch to `None`, however, as I've never used this class, just written the bindings

File: vital/types/metadata.h
Comment (daniel-riehm) [reply vital/types/metadata.h:213]: Looking back on it now that I understand things better, I don't think there is sufficient reason to change it. I'll restore the original version.

File: python/kwiver/vital/tests/test_metadata.py
Comment (russelldj) [reply python/kwiver/vital/tests/test_metadata.py]: I'm slightly lost here. Is `unknown_t` the class returned by a query which doesn't contain that metadata item? It seems like in the past we've compared the tag to the `VITAL_METADATA_UNKNOWN` tag. So if this were `None`, would that break?

Comment (daniel-riehm) [reply python/kwiver/vital/tests/test_metadata.py]: No, it shouldn't break. 
The previous behavior of `inst.data` was to return the integer value `0`. Now, it returns `None`, but the tag has not changed. 
If you think of the metadata like a `{ tag: data }` dictionary, the entry was `{ VITAL_METADATA_UNKNOWN: 0 }`; now it is `{ VITAL_METADATA_UNKNOWN: None }`. In my opinion, this makes more sense; an unknown tag doesn't have an integer value, it has no value (`None`).
On the C++ end, Python's built-in `None` type is replaced here with a dummy `unknown_t` class, the point of which is that, unlike the integer `0`, it cannot be accidentally interpreted as a legitimate value of any sort. Right now, if you do `metadata.find(SOME_TAG).as_double()` and the tag is not found, the `as_double()` will silently return `0.0`, since `int` can be casted to `double`. This is a problem because the program will continue on like the metadata field actually had the value `0.0`. With these new changes, `as_double()` will instead throw a runtime error, since `unknown_t` cannot be casted to `double`, alerting you to the bug.

Comment (russelldj) [reply python/kwiver/vital/tests/test_metadata.py]: Thanks for the through response, it makes a lot more sense now. I agree that `None` is a very reasonable value.

Comment (tao558) [reply python/kwiver/vital/tests/test_metadata.py]: I think it would also be reasonable to change what `type()` returns as well.

Comment (daniel-riehm) [reply python/kwiver/vital/tests/test_metadata.py]: @tao558 What would you suggest? Currently it returns the fully-qualified name of the C++ type, which seems to be standard elsewhere in the file (e.g. `kwiver::vital::geo_point`)

Comment (tao558) [reply python/kwiver/vital/tests/test_metadata.py]: Right thats fine on the C++ side, but that type on the Python side is now `type(None)`, since we changed data to be None. We're still returning `typeid(unknown_t)` in the Python bindings even though we've changed the data to `None`

Comment (daniel-riehm) [reply python/kwiver/vital/tests/test_metadata.py]: To the string literal `"none"`, then? Or `"py::none"` or `"None"`?

Comment (daniel-riehm) [reply python/kwiver/vital/tests/test_metadata.py]: That gives me a `'py::type' has not been declared` error. I looked in the pybind11 repo and `git blame` says that `py::type` class was only added in the past year. I don't think Fletch's version of `pybind11` has that feature.

File: python/kwiver/vital/types/metadata.cxx
Comment (tao558) [inline python/kwiver/vital/types/metadata.cxx:119]: Running `inst.type is None` will return `False` with this check... Try using the older version: `get_type()`. Pybind11 should have functionality for this somewhere if this doesnt work

**Review by tao558 — CHANGES_REQUESTED** [review]

File: python/kwiver/vital/types/metadata.cxx
Comment (daniel-riehm) [reply python/kwiver/vital/types/metadata.cxx:119]: You want `type()` to return `None`? Shouldn't we just return `py::none()` then?

Comment (tao558) [reply python/kwiver/vital/types/metadata.cxx]: No, we want `type()` to return `NoneType` here. If you run `type(None)` from the Python interpreter, you get `<class 'NoneType'>`. Right now, `data` returns `None`, so we want `type` to return the equivalent of running `type(None)` from the interpreter. Doing something like `return py::none().get_type()` might do the trick.

Comment (daniel-riehm) [reply python/kwiver/vital/types/metadata.cxx]: I changed it to `py::none().get_type()`.

**Comment by daniel-riehm** [comment]
@tao558 Is this good to land now?

**Comment by tao558** [comment]
> @tao558 Is this good to land now?

One last thing. Add an extra check [here](https://github.com/Kitware/kwiver/pull/1351/files#diff-dbbd16b875a66de72a965a0e728f938aab2b34ed8ccf4c5eaee7233ffba56fd2R257) for `None`. Something like:
```
elif prop_info.data is None:
    nt.ok_(inst.data is None)
```

After that I'll approve. I'm leaving the C++ changes to those who have already reviewed, although I did take a glance at them.

**Review by tao558 — APPROVED** [review]
Pending passing CI

**Pull request merged**
